File: python/local_context_engine/compression.py
"""Compression strategies for fitting more context into a finite window.

The prototype uses deterministic local compression:
- full inclusion for high-priority chunks
- extractive sentence compression for chunks that only partially fit
- a compact "omitted memory" note so the model knows what was left out
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Optional, Tuple

from .models import RetrievedItem, estimate_tokens
from .text_utils import split_sentences, term_frequencies, tokenize


@dataclass
class PackedItem:
    item: RetrievedItem
    content: str
    tokens: int
    compressed: bool = False


@dataclass
class CompressionResult:
    packed: List[PackedItem]
    omitted_count: int = 0
    omitted_tokens: int = 0
    omitted_ids: List[str] = None

    def __post_init__(self) -> None:
        if self.omitted_ids is None:
            self.omitted_ids = []


def compress_items(
    items: Iterable[RetrievedItem],
    budget_tokens: int,
    query: str = "",
    min_chunk_tokens: int = 24,
) -> CompressionResult:
    """Greedily pack items into a token budget, compressing the overflow."""
    ordered = sorted(items, key=lambda item: item.score, reverse=True)
    packed: List[PackedItem] = []
    omitted_ids: List[str] = []
    omitted_tokens = 0
    used = 0
    for item in ordered:
        content = item.chunk.content
        tokens = item.chunk.tokens or estimate_tokens(content)
        if used + tokens <= budget_tokens:
            packed.append(PackedItem(item=item, content=content, tokens=tokens, compressed=False))
            used += tokens
            continue

        remaining = budget_tokens - used
        if remaining >= min_chunk_tokens:
            compressed, compressed_tokens = compress_chunk_to_fit(
                content, remaining, query=query
            )
            if compressed_tokens > 0:
                packed.append(
                    PackedItem(item=item, content=compressed, tokens=compressed_tokens, compressed=True)
                )
                used += compressed_tokens
                continue

        omitted_ids.append(item.chunk.id)
        omitted_tokens += tokens

    return CompressionResult(
        packed=packed,
        omitted_count=len(omitted_ids),
        omitted_tokens=omitted_tokens,
        omitted_ids=omitted_ids,
    )


def compress_chunk_to_fit(content: str, max_tokens: int, query: str = "") -> Tuple[str, int]:
    """Compress a single chunk to fit max_tokens.

    First tries extractive sentence ranking; if that is still too big, falls
    back to hard truncation at the token boundary.
    """
    if max_tokens <= 0:
        return "", 0
    if estimate_tokens(content) <= max_tokens:
        return content, estimate_tokens(content)

    extractive = _extractive_compress(content, max_tokens, query)
    if extractive:
        return extractive, estimate_tokens(extractive)

    return truncate_text(content, max_tokens)


def truncate_text(content: str, max_tokens: int) -> Tuple[str, int]:
    """Truncate text to an approximate token budget while keeping whole lines."""
    if max_tokens <= 0:
        return "", 0
    if estimate_tokens(content) <= max_tokens:
        return content, estimate_tokens(content)
    parts = content.splitlines(keepends=True)
    out: List[str] = []
    used = 0
    for part in parts:
        part_tokens = estimate_tokens(part)
        if used + part_tokens > max_tokens:
            # If a single long line still exceeds the remaining budget, cut it.
            remaining = max_tokens - used
            if remaining > 8 and part:
                cut = _cut_by_tokens(part, remaining)
                out.append(cut)
                used += estimate_tokens(cut)
            break
        out.append(part)
        used += part_tokens
    text = "".join(out).rstrip()
    if not text:
        # last resort: hard character cut
        text = _cut_by_tokens(content, max_tokens)
    return text, estimate_tokens(text)


def _extractive_compress(content: str, max_tokens: int, query: str = "") -> str:
    """Pick the most informative sentences under a token budget."""
    sentences = split_sentences(content)
    if len(sentences) <= 1:
        return ""
    query_freqs = term_frequencies(tokenize(query)) if query else {}
    scored = []
    for idx, sentence in enumerate(sentences):
        freqs = term_frequencies(tokenize(sentence))
        overlap = sum(freqs.get(t, 0) * qf for t, qf in query_freqs.items())
        length_bonus = min(1.0, len(sentence) / 200.0)
        position_bonus = 1.0 if idx == 0 else 0.6
        score = overlap + length_bonus * 0.5 + position_bonus
        scored.append((score, idx, sentence))
    scored.sort(key=lambda x: x[0], reverse=True)
    chosen: List[Tuple[int, str]] = []
    used = 0
    for _, idx, sentence in scored:
        sentence_tokens = estimate_tokens(sentence)
        if used + sentence_tokens > max_tokens:
            continue
        chosen.append((idx, sentence))
        used += sentence_tokens
    if not chosen:
        return ""
    chosen.sort(key=lambda x: x[0])
    return "\n".join(s for _, s in chosen)


def build_omitted_note(omitted_count: int, omitted_tokens: int, max_tokens: int = 80) -> str:
    """A small note that tells the model older/less relevant context was omitted."""
    if omitted_count <= 0:
        return ""
    note = (
        f"[Local Context Engine] {omitted_count} lower-priority item(s) "
        f"(~{omitted_tokens} tokens) were omitted to stay within the context budget. "
        "Ask the user or use retrieval if any omitted detail is needed."
    )
    if estimate_tokens(note) > max_tokens:
        note = (
            f"[Local Context Engine] {omitted_count} lower-priority item(s) "
            f"(~{omitted_tokens} tokens) omitted for budget."
        )
    return note


def _cut_by_tokens(text: str, max_tokens: int) -> str:
    """Cut text by approximate tokens, preserving CJK characters."""
    if max_tokens <= 0:
        return ""
    out = []
    used = 0
    for ch in text:
        cost = 1 if "\u4e00" <= ch <= "\u9fff" else 0.25
        if used + cost > max_tokens:
            break
        out.append(ch)
        used += cost
    return "".join(out).rstrip()

File: engine/compression.js
// Compression strategies for fitting more context into a finite window.

import { estimateTokens } from "./models.js";
import { splitSentences, termFrequencies, tokenize } from "./text_utils.js";

export class PackedItem {
  constructor({ item, content, tokens, compressed = false }) {
    this.item = item;
    this.content = content;
    this.tokens = tokens;
    this.compressed = compressed;
  }
}

export class CompressionResult {
  constructor({ packed = [], omittedCount = 0, omittedTokens = 0, omittedIds = [] } = {}) {
    this.packed = packed;
    this.omittedCount = omittedCount;
    this.omittedTokens = omittedTokens;
    this.omittedIds = omittedIds;
  }
}

export function compressItems(items, budgetTokens, query = "", minChunkTokens = 24) {
  const ordered = [...items].sort((a, b) => b.score - a.score);
  const packed = [];
  const omittedIds = [];
  let omittedTokens = 0;
  let used = 0;

  for (const item of ordered) {
    const content = item.chunk.content;
    const tokens = item.chunk.tokens || estimateTokens(content);
    if (used + tokens <= budgetTokens) {
      packed.push(new PackedItem({ item, content, tokens, compressed: false }));
      used += tokens;
      continue;
    }

    const remaining = budgetTokens - used;
    if (remaining >= minChunkTokens) {
      const [compressed, compressedTokens] = compressChunkToFit(content, remaining, query);
      if (compressedTokens > 0) {
        packed.push(new PackedItem({ item, content: compressed, tokens: compressedTokens, compressed: true }));
        used += compressedTokens;
        continue;
      }
    }

    omittedIds.push(item.chunk.id);
    omittedTokens += tokens;
  }

  return new CompressionResult({ packed, omittedCount: omittedIds.length, omittedTokens, omittedIds });
}

export async function compressItemsAsync(items, budgetTokens, query = "", { minChunkTokens = 24, summarizer = null } = {}) {
  const ordered = [...items].sort((a, b) => b.score - a.score);
  const packed = [];
  const omittedIds = [];
  let omittedTokens = 0;
  let used = 0;

  for (const item of ordered) {
    const content = item.chunk.content;
    const tokens = item.chunk.tokens || estimateTokens(content);
    if (used + tokens <= budgetTokens) {
      packed.push(new PackedItem({ item, content, tokens, compressed: false }));
      used += tokens;
      continue;
    }

    const remaining = budgetTokens - used;
    if (remaining >= minChunkTokens) {
      let compressed = "";
      if (summarizer) {
        compressed = await summarizer(content, remaining);
      }
      if (!compressed) {
        [compressed] = compressChunkToFit(content, remaining, query);
      }
      const compressedTokens = estimateTokens(compressed);
      if (compressedTokens > 0) {
        packed.push(new PackedItem({ item, content: compressed, tokens: compressedTokens, compressed: true }));
        used += compressedTokens;
        continue;
      }
    }

    omittedIds.push(item.chunk.id);
    omittedTokens += tokens;
  }

  return new CompressionResult({ packed, omittedCount: omittedIds.length, omittedTokens, omittedIds });
}

export function compressChunkToFit(content, maxTokens, query = "") {
  if (maxTokens <= 0) return ["", 0];
  if (estimateTokens(content) <= maxTokens) return [content, estimateTokens(content)];

  const extractive = extractiveCompress(content, maxTokens, query);
  if (extractive) return [extractive, estimateTokens(extractive)];

  return truncateText(content, maxTokens);
}

export function truncateText(content, maxTokens) {
  if (maxTokens <= 0) return ["", 0];
  if (estimateTokens(content) <= maxTokens) return [content, estimateTokens(content)];
  const parts = content.split(/(?<=\n)/);
  const out = [];
  let used = 0;
  for (const part of parts) {
    const partTokens = estimateTokens(part);
    if (used + partTokens > maxTokens) {
      const remaining = maxTokens - used;
      if (remaining > 8 && part) {
        const cut = cutByTokens(part, remaining);
        out.push(cut);
        used += estimateTokens(cut);
      }
      break;
    }
    out.push(part);
    used += partTokens;
  }
  let text = out.join("").trimEnd();
  if (!text) text = cutByTokens(content, maxTokens);
  return [text, estimateTokens(text)];
}

function extractiveCompress(content, maxTokens, query = "") {
  const sentences = splitSentences(content);
  if (sentences.length <= 1) return "";
  const queryFreqs = query ? termFrequencies(tokenize(query)) : {};
  const scored = [];
  for (let idx = 0; idx < sentences.length; idx += 1) {
    const sentence = sentences[idx];
    const freqs = termFrequencies(tokenize(sentence));
    let overlap = 0;
    for (const [term, qf] of Object.entries(queryFreqs)) {
      overlap += (freqs[term] || 0) * qf;
    }
    const lengthBonus = Math.min(1, sentence.length / 200);
    const positionBonus = idx === 0 ? 1 : 0.6;
    const score = overlap + lengthBonus * 0.5 + positionBonus;
    scored.push({ score, idx, sentence });
  }
  scored.sort((a, b) => b.score - a.score);
  const chosen = [];
  let used = 0;
  for (const { idx, sentence } of scored) {
    const sentenceTokens = estimateTokens(sentence);
    if (used + sentenceTokens > maxTokens) continue;
    chosen.push({ idx, sentence });
    used += sentenceTokens;
  }
  if (chosen.length === 0) return "";
  chosen.sort((a, b) => a.idx - b.idx);
  return chosen.map((entry) => entry.sentence).join("\n");
}

export function buildOmittedNote(omittedCount, omittedTokens, maxTokens = 80) {
  if (omittedCount <= 0) return "";
  let note =
    `[Local Context Engine] ${omittedCount} lower-priority item(s) ` +
    `(~${omittedTokens} tokens) were omitted to stay within the context budget. ` +
    "Ask the user or use retrieval if any omitted detail is needed.";
  if (estimateTokens(note) > maxTokens) {
    note =
      `[Local Context Engine] ${omittedCount} lower-priority item(s) ` +
      `(~${omittedTokens} tokens) omitted for budget.`;
  }
  return note;
}

function cutByTokens(text, maxTokens) {
  if (maxTokens <= 0) return "";
  const out = [];
  let used = 0;
  for (const ch of String(text)) {
    const cost = ch >= "\u4e00" && ch <= "\u9fff" ? 1 : 0.25;
    if (used + cost > maxTokens) break;
    out.push(ch);
    used += cost;
  }
  return out.join("").trimEnd();
}
